### PR TITLE
Web Component Server Side Rendering (SSR)

### PR DESCRIPTION
**Pull request by sghoweri**
## Jira
http://vjira2:8080/browse/BDS-749?filter=-1

## Summary
Adds the capability to server-side pre-render Bolt web components via a new new experimental `@bolt/ssr-server` package. 

## Details
- Updated the `@bolt/components-button` templating logic to conditionally adjust what specific markup gets used. Server-side rendered buttons now output virtually the same markup as the button’s Twig template while client-side rendered buttons retain the same markup as before.
- Updated the `<bolt-text>`, `<bolt-button>`, `<replace-with-children>`, and `<replace-with-grandchildren>` templating logic to better handle client-side hydration when server-side rendered.
- The Bolt build tools have been updated with a new under the hood `mode` config which adjusts some of the the build process‘s behavior (ex. no CSS file’s are compiled to speed up the build times) + provides this contextual info to the the components being compiled and rendered.
    - This means that the client-side JS differs slightly from the server-side JS and is generated as a bolt-global.server.js bundle 
    - This MIGHT ultimately not be necessary however further testing and dev work would be needed to confirm whether or not this is feasible and practical.
- Added an E2E integration test with Nightwatch.js to confirm SSR is working as expected in Pattern Lab
- Added a handful of component examples to test the SSR functionality is working as expected:
    - Button component (ex. 
https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-button--button-server-side-rendering--wo-shadow-dom/02-components-button--button-server-side-rendering--wo-shadow-dom.html, among a few others)
    - Text component 
    - Icon component
- Added a new `cleanAll` optional param to the internal `@bolt/build-tools` clean task -- wired up to be used by Jest to ensure tests are being run in a sanitized environment for improved test consistency
- Refactored and relocated the previous top level `servers` folder to now live in the `packages/servers` folder
- Fixed a small issue encountered with the `@bolt/build-tools` that had been preventing Pattern Lab and the Static Site from getting manually compiled via the CLI (unless specified in the `.boltrc` config)
- Reorganized a few files in the `@bolt/core-php` package to be structured more consistently
- Added the missing `fileExists` and `inlineFile` Twig functions to Bolt Core (which should have been added previously)
- ~~Addressed an issue with the Jest visual regression testing logic which throws an error if a `NOW_TOKEN` env token isn't available locally~~
- Updated the `Dockerfile` to account for the new express server location in the codebase 
- Fixed an unrelated issue with our component Jest test async `describe` calls [throwing Jest warnings](https://travis-ci.com/bolt-design-system/bolt/jobs/184204759)
- Updated several docs site templates as a workaround to fix unrelated upstream Twig updates that had been [breaking the build](https://travis-ci.com/bolt-design-system/bolt/jobs/184204760)

Note:
- While this implementation does include a twig filter, function, and tag integrations to use server-side rendering in Twig, the `{% ssr %}` tag isn't ready for prime time (yet) due to the Twig environment's `context` not being passed along. Perhaps @remydenton has some experience with this...

<details>
  <summary><strong>Usage</strong></summary>
This brand new SSR package currently provides three different interfaces to pre-render web components:

### Option 1. via the CLI (super helpful for testing). 

This Node.js CLI currently accepts two arguments:
    1. The string of HTML to pre-render
    2. Boolean to conditionally enable/disable syntax highlighting (which defaults to `true` when using the CLI directly, `false` when being used for PHP rendering)

```
$ node packages/servers/ssr-server/cli.js '<bolt-button color="secondary">SSR Test</bolt-button>'
```

This returns back the pre-rendered, hydration-friendly markup (with or without syntax highlighting)

image

### Option 2. via Twig filter / function
There’s also two ways to use this SSR functionality in Twig via the new `bolt_ssr` Twig filter and `bolt_ssr` Twig function. 

Both of these also use the Node.js CLI interface under the hood with the added bonus of returning back the original pre-rendered HTML as a fallback.
```
{% filter bolt_ssr %}
  <bolt-button color=“secondary”>
    SSR Test
  </bolt-button>
{% endfilter %}
```

### Option 3. via JS using the SSR library itself (experimental)

This SSR functionality can even be used in other JS files via the `renderToString` method on the `@bolt/ssr-server` package.

> Note: currently this method currently requires the developer to manually call the `shutDownSSRServer` method afterwards — hopefully something we can automate in the next iteration.
</details>

<details>
  <summary><strong>Misc Notes</strong></summary>
<br>  
- The first couple of implementations of this were powered by a Google Puppeteer instance however this latest iteration managed to swap out Puppeteer for a lighter weight JSDOM implementation that’s been lightly polyfilled! 🎉
- SSR compile times are averaging around 1 to 1.3 seconds (when no Webpack build is required). Quick enough for a larger POC but not quick enough to globally roll this out without expensive compile times on the server   
</details>

<details>
  <summary><strong>Gotchas / Next Steps</strong></summary>

- **Auto-wire Components Waiting To Load**: we should figure out how we want to create a list of the custom element tags that need to be defined before the SSR logic kicks off: https://github.com/bolt-design-system/bolt/blob/cd72b43574e7d33810c80147edc8cb1613048a38/packages/servers/ssr-server/renderer.js#L42 
- **Cache Busting**: Further work needed to get cache busting added to the SSR build process
- **Async Rendering**: Look into parallelizing / making requests more async
- **Alt Interface for PHP Integtation?** We should look into using web sockets or even using a vanilla server request / response interface.
- **Caching**: making the same request to the same component with the same data on the same version of Bolt should pull straight from memory to speed up build times.
- **Compare SSR Perf**: Further research into performance optimizations -- including doing a benchmark between say, SSR for a Vue.js component compared to our current implementation
- **Test CKEditor Integration**: Test integrating CKEditor example content with this new SSR functionality
- **Combine Webpack Builds?** Look into (potentially) combining the client-side vs server-side webpack builds to eliminate having to compile some of the same components twice
- **Research Further Optimizing Client-side Perf**: Lighter weight JS load in the client = faster SSR response times
- **Combine Multiple Servers**: Out of simplicity, we currently have 3 different Express servers -- the default Express server, the testing server when running Jest tests, and now this new SSR server. At a minimum I could see two of these getting merged together into a single server instance (if not merging together all three)
- **Handle Pre-rendered HTML**: the current implementation doesn't know if the markup provided (or pieces of the markup provided) has been pre-rendered which could lead to unintended rendering quirks and/or extra unnecessary processing 
- **Fragment Cache?** If a button component was being pre-rendered but contained an already pre-rendered icon component, how might we handle that?

</details>

<details>
  <summary><strong>Performance Impact</strong></summary>

**Notes:**
1. This is based on the current implementation which will reuses the `bolt-global.server.js` if a Webpack build had already been done -- this speeds up the initial SSR request.
2. CLI benchmarks were are based on this `bolt-global.server.js` file already existing
3. The current implementation is synchronous -- only one request at a time and PHP waits for the request to complete before continuing.

<hr>

### CLI Performance (per compile):
- Without formatting HTML or syntax highlighting: 1.08s average (or 0.05s faster on average than with Prettier formatting)
- With highlighting + prettier: 1.23s average
- Without syntax highlighting, with HTML formatting: 1.13s average

### Pattern Lab Compile Times (Prod) with SSR enabled (Select Usage)
- Using it in a handful of places (~7 places): ~21.97s total compile time
- Without SSR enabled: ~13.s total compile time

### Global Usage (Tests SSR `<bolt-button>` Globally)
- Pre-rendering all `<bolt-button>` components in the static site generator: ~1 minute, 40s compile time (vs 4s compile time normally)
- Pre-rendering all `<bolt-button>` components in Pattern Lab: 11 minutes, 27.7s minutes (vs 13.2s normally)
</details>

## How to test
- [x] Confirm Travis Build working as expected
- [x] Test running build locally to confirm local compile times of PL are acceptable
- [x] Review the handful of component examples that are wired up to use this new SSR functionality (a few `<bolt-button>` demos, a `<bolt-text>` demo, and a few `<bolt-icon>` demos)

### CODE REVIEW

File: docs-site/src/pages/pattern-lab/_patterns/02-components/button/70-server-side-rendering--twig-only.twig [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{% filter bolt_ssr %}
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/02-components/button/70-server-side-rendering--twig-only.twig]: Hmm.... The whitespace I get (that could be an issue with Prettier trying to clean up the HTML before returning it) however I'm not sure I'm understanding the problem with the nested `<replace-with-children>` "issue" you're mentioning....


OH WAIT. Are you talking about this extra `<replace-with-children class="c-bolt-button__item">` that's wrapping around the content?
```
<bolt-button
  align="center"
  color="primary"
  size="medium"
  transform="none"
  border-radius="regular"
>
  <button
    class=" c-bolt-button c-bolt-button--center c-bolt-button--primary c-bolt-button--medium c-bolt-button--none c-bolt-button--border-radius-regular"
  >
    <replace-with-children class="c-bolt-button__item">
      <replace-with-children
        class="c-bolt-button__icon is-empty"
      ></replace-with-children>

      <replace-with-children class="c-bolt-button__item "
        >SSR via pure Twig</replace-with-children
      >
      <replace-with-children
        class="c-bolt-button__icon is-empty"
      ></replace-with-children>
    </replace-with-children>
  </button>
</bolt-button>
```

File: docs-site/src/pages/pattern-lab/_patterns/02-components/button/-button-server-side-rendering--wo-shadow-dom.twig [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{# test component used for Nightwatch.js test that checks if server-side pre-rendering is working in Twig as expected #}
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/02-components/button/-button-server-side-rendering--wo-shadow-dom.twig]: Updated and added some extra comments to explain what that demo is used for.

**Comment by sghoweri** [comment]
> * text/30-server-side-rendering--wc-only.twig
> * icon/30-server-side-rendering--twig-only.twig
> * icon/30-server-side-rendering--wc-only.twig

I've updated all 3 of these component demos to make sure they show up correctly in Pattern Lab's navigation. Taking a look at the two other `<bolt-text>` components to see what's going on

**Comment by sghoweri** [comment]
> These pages now appear to be blank, regression?:
> 
> * [Text icon variations](https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-text--text-icon-variations/02-components-text--text-icon-variations.html)
> * [Text size and weight variations](https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-text-25-text-size-and-weight-variations/02-components-text-25-text-size-and-weight-variations.html)

I've updated the text w/ icon variations page to hide it completely from the menu until we're ready to bring it back to life (post `<bolt-text>` cleanup most likely).

As for the other text size / weight variations page -- that's been broken for a while due to a few schema case changes from 8 months ago (so totally unrelated). In any case, I've also updated that PL demo to get that back up and running.

**Comment by sghoweri** [comment]
> Thoughts in general:
> 
> * I would definitely like to see more demos/tests handling pre-rendered/fragments of pre-rendered elements. If there are bugs lurking there, let's get out ahead of them.
Totally agree. The plan is for us to continue to throughly test (plus continue to work on SSR performance) before we move to the next level of testing / iteration.

Also I’ve been thinking that it probably makes the most sense for us to slowly roll out support at the component level once things have been tested / updated accordingly.

For example: button required a couple small updates, text required even fewer, and icon required no updates whatsoever.

> * It seems like performance/speed is going to be a major consideration for making the switch to full SSR. Is SSR in Vue much faster?

It depends. Super tiny simple SSR Vue demos are really quick BUT:
1. They often  require a Node server to be running continuously in the background
2. They have poor support for any libraries or code that expect / require to be running in an actual browser environment (ie. `window is undefined`).

TLDR: I don’t know yet. Performance and reliability are top priorities for this but super steep limitations could also limit the value added.

For what it’s worth, JSDOM is nearly universally the goto “fast” method for handing this so I’m curious to see how fast this current setup could run with a fully optimized JS build. 

> * I'm not crazy about the duplication `isServer` [requires in the JS](https://github.com/bolt-design-system/bolt/pull/1097/files#diff-ad8093a42b9e5fc5b748d20b6e11b078R78). On the other hand, if it means dropping Twig at some point, that's still an improvement. Might be a pattern we can improve upon in the future.

Dev huddle discussion? I’m not a huge fan either BUT this would be our replacement for Twig when we can’t just delete a component’s inner HTML when the JS boots up.

Some components won’t require this but others with more advanced hydration / no js functionality would need something along these lines.

> * Don't totally understand what those SSR PHP files are doing. Looks like some voodoo. Possible dev huddle topic?

Sure - dev huddle chat. 😉

File: docs-site/src/pages/pattern-lab/_patterns/02-components/button/70-server-side-rendering--twig-only.twig (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielamorse) [reply docs-site/src/pages/pattern-lab/_patterns/02-components/button/70-server-side-rendering--twig-only.twig]: Yes, exactly. In the other examples there's no wrapping `replace-with-children`.

File: docs-site/src/pages/pattern-lab/_patterns/02-components/button/-button-server-side-rendering--wo-shadow-dom.twig (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielamorse) [reply docs-site/src/pages/pattern-lab/_patterns/02-components/button/-button-server-side-rendering--wo-shadow-dom.twig]: Look good!

**Comment by danielamorse** [comment]
@sghoweri reviewed again and just have a couple more comments:
 - In the [icon example](https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-icon-30-icon-ssr-example--twig/02-components-icon-30-icon-ssr-example--twig.html), it looks like the code snippets aren't rendering properly.
image

- In the [text example](https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-text-30-text-ssr-example/02-components-text-30-text-ssr-example.html), there is a JS error
image

Neither of these would be show stoppers for approving the PR. At this point just waiting on the `replace-with-children` issue, and even that is a minor issue. If you want to handle it in a follow-up PR, just let me know and I'll approve.

**Comment by sghoweri** [comment]
> @sghoweri reviewed again and just have a couple more comments:
> 
> * In the [icon example](https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-icon-30-icon-ssr-example--twig/02-components-icon-30-icon-ssr-example--twig.html), it looks like the code snippets aren't rendering properly.
>   image
> * In the [text example](https://feature-server-side-rendering.boltdesignsystem.com/pattern-lab/patterns/02-components-text-30-text-ssr-example/02-components-text-30-text-ssr-example.html), there is a JS error
>   image
> 
> Neither of these would be show stoppers for approving the PR. At this point just waiting on the `replace-with-children` issue, and even that is a minor issue. If you want to handle it in a follow-up PR, just let me know and I'll approve.

Thanks @danielamorse. Looks like the button component's main JS file got renamed recently so I folded the couple of template logic changes in and removed the old `button.standalone.js` file. I THINK that should solve part of these last couple of things you're flagging here. 

On top of that I fixed two button demos' icons that were missing slots + updated all the SSR demos to make sure all web components being talked about in the demos don't accidentally render 🤦‍♂️  -- that's likely the main issue you're seeing in the first screenshot above (no icon `name` prop defined = no SVG renders).
